articles/user-guide/host-programs.md

```diff
@@ -0,0 +1,455 @@
+
+### Operation defined in a Q# file
+
+The operation is precisely what's called and run by Q#.
+However, it requires a few more additions to comprise a full `*.qs` Q# file.
+
+All Q# types and callables are defined within *namespaces*, which the compiler can then open individually 
+
+Firstly, the [`H`](xref:microsoft.quantum.intrinsic.h) and [`MResetZ`](xref:microsoft.quantum.measurement.mresetz) actually belong to the [Q# Standard Libraries](xref:microsoft.quantum.qsharplibintro).
+To be available for use, their respective *namespaces* need to be opened.
+All Q# callables and types (both those you define and those intrinsic to the language) are defined within namespaces.
+To access them, the compiler needs to be made aware of which namespaces to open.
+
+So, the full Q# file containing our operation consists of defining our own namespace, opening the namespaces for the operations our operation uses, and then our operation:
+
+```qsharp
+namespace NamespaceName {
+    open Microsoft.Quantum.Intrinsic;     // for the H operation
+    open Microsoft.Quantum.Measurement;   // for MResetZ
+
+    operation MeasureSuperposition() : Result {
+        using (q = Qubit()) { 
+            H(q);
+            return MResetZ(q);
+        }
+    }
+}
+```
+
+### Execution of a Q# program
+
+Now the general execution model of a Q# program becomes clear.
+
+<br/>
+<img src="./images/general_execution_model.png" alt="Q# program execution diagram" width="400">
+
+Firstly, the specific operation to be executed has access to any other callables and types defined in the same namespace.
+Thanks to `open` statements, it can also access those from any of the [Q# libraries](xref:microsoft.quantum.libraries). 
+
+The operation itself is then executed on a *[target machine](xref:microsoft.quantum.machines)*.
+In the future, one such target machine will be a real quantum computer, but there are currently multiple simulators available, each with a particular use.
+For our purposes here, the most useful target machine is instance of the full-state simulator, `QuantumSimulator`, which calculates the program's behavior as if it were being executed on a noise-free quantum computer.
+
+So far, we've described what happens when a specific Q# operation is being executed.
+Regardless of whether Q# is used in a standalone application or with a host program, this general process is more or less the same---hence the QDK's flexibility.
+The differences between the methods therefore reveal themselves in *how* that Q# operation is called to be executed, and in what manner any results are returned.
+
+First, we discuss how this done with the Q# standalone application from the command line, and then proceed to using Python and C# host programs.
+We reserve the standalone application of Q# Jupyter Notebooks for last, because unlike the first three, it does not actually require a Q# file.
```

Comment:
In other places, we've had the two Q#-only workflows first and then the two interop workflows after that. Strictly speaking, Python also doesn't require Q# files, as you can use the `qsharp.compile` function to compile on the fly; that's definitely not a typical workflow, though.

Reply:
That's another thing I felt could use some discussion/input. It did feel odd placing it at the end, but at the same time, the (typical) procedural workflow of the two interops seem to flow so naturally from that of the command-line (i.e. given the Q# file, everything we were doing with `EntryPoint`, we simply have the host program call). Given all those similarities, it felt like it might be a little jarring to jump to the notably different Jupyter NB situation (e.g. for now, no arguments), only to then return to the host program flow afterward. 

I don't think that would be an issue if this page were to be more of a quick-reference for each workflow, but what I had in mind was that many people may be coming here even _before_ choosing a workflow. Hence they would be more likely to read through it as a whole piece, getting an idea of how each works to inform their decision. Definitely something I'll bring up to more people in tomorrow's call